### DISCUSSION TITLE
IResourceHandler proper response

### DISCUSSION THREAD

**Discussion by crisleiria** [cefsharp/CefSharp#3423] [Q&A]
I am using CafSharp wpf some time now, and now its time to update do latest versions.

I created an custom link and implemented with IResourceHandler. This custom link open file explorer on an specific network path.
This was working with no problems.

After update to latest version (87.1.132), when I click on my custom link, application opens explorer on the correct network path but also open File Download dialog to download file with no data.. What is the cause of this?
Do I need to specify some sort of response headers?

I have implemented IResourceHandler.ProcessRequest(IRequest request, ICallback callback) for this.

Please advice, thanks in advance.
Cristovao Oliveira

**Comment by amaitland** [comment]
> After update to latest version (87.1.132)

Does the problem reproduce version 88.2.90?

Without a detailed example I cannot even speculate what's going on here.

**Comment by crisleiria** [comment]
I have implemented a custom Interceptor class for links:

public sealed class ShellSchemeHandler : ISchemeHandlerFactory, IResourceHandler



`




        bool IResourceHandler.Open(IRequest request, out bool handleRequest, ICallback callback)
        {
            try
            {
                // Make sure URL matches.
                var match = ShellRegex.Match(request.Url);
                if (!match.Success)
                {
                    handleRequest = false;
                    return false;
                }
                var targetPath = match.Groups[1].Value.ToUpperInvariant();
                targetPath = targetPath.Replace('/', '\\');


                // Tries to open the file/folder.
                Process.Start("explorer.exe", $"\\\\{targetPath}");
                callback.Continue();
               
            }
            catch (Exception exception)
            {
                MessageBox.Show($"An exception '{exception.GetType()}' was thrown with the following message: {exception.Message}.", "Error on executing operation", MessageBoxButton.OK, MessageBoxImage.Error);
            }

            handleRequest = true;
            return true;
        }

`


![Untitled](https://user-images.githubusercontent.com/9845290/110177480-f32d6080-7dfc-11eb-957a-39faf38350b6.png)

When I click on custom link (Green folder), if I have Regex Match I want to open explorer on custom path. On version 87 and 88 it opens explorer but also File Save Dialog on the same moment. I notice changes on IResourceHandler interface... problaly I need to change some sort of response or change something on callback object.

This appens on versions 87 and 88.

**Reply by amaitland** [reply]
What version are you upgrading from? 

> I notice changes on IResourceHandler interface

What changes are you referring to? Other than typos the interface hasn't changed in 18mths. https://github.com/cefsharp/CefSharp/commits/8afe0dfca6fd3d2ae90b08941fe89828e700174c/CefSharp/Handler/IResourceHandler.cs

What you are seeing is likely the result of a change in CEF/Chromium. 

The API doc is at http://cefsharp.github.io/api/88.2.x/html/M_CefSharp_IResourceHandler_Open.htm

The CEF documentation is at https://github.com/chromiumembedded/cef/blob/cae194f493328a9890b27533ce1dc661e0d4b042/include/cef_resource_handler.h#L90

If your intention is to cancel the request then I believe your code is incorrect. In future please provide a complete code example, just a snippet doesn't paint the whole picture.

**Reply by crisleiria** [reply]
I come from version 73... 

I was implementing ProcessRequest and I see that is deprecated now.

I pasted the same code from ProcessRequest to Open method basically.

I will check with more detail your references links and test. I will reply with my findings (only next week ;)).

Thank you for your support @amaitland

**Reply by amaitland** [reply]
> I come from version 73...

In future please provide this information when asking a question, it makes a big difference, I'm assuming you are upgrading from the previous version when you are upgrading from a very very old version.

There are a huge number of changes at the `CEF/Chromium` level, please make sure you read #2743


> I pasted the same code from ProcessRequest to Open method basically.

This is unlikely to work.

If you require further assistance please provide the an copy of your original `IResourcerHandler` implementation

**Reply by crisleiria** [reply]
Dear @amaitland 

Thank you for your help. Sorry for not placing version info.

Well it opens explorer but it also opens FileSave Dialog... So I believe the request is still be handled on some events after.
I also notice IResourceHandler.Read is called.

Here is the working version class of ShellSchemeHandler from version 73

https://drive.google.com/file/d/1Bf7qcWcbikFFZNVUZIksXfkQQ8jgjzOX/view?usp=sharing

Basically what I have done was ProcessRequest to Open method basically.
If possible please advise what I can do here.

**Reply by amaitland** [reply]
> Here is the working version class of ShellSchemeHandler from version 73

Please use a code sharing service like GitHub gist

https://gist.github.com/

**Reply by crisleiria** [reply]
https://gist.github.com/crisleiria/33fa169af77c6f435c1382a2ff9e825e

**Reply by amaitland** [reply]
See https://github.com/cefsharp/CefSharp/wiki/General-Usage#protocol-execution for the approach I'd recommend. 

Using a custom scheme for this use case is not recommend now. Canceling the navigation in OnBeforeBrowse is the simplest option.

**Reply by crisleiria** [reply]
Dear @amaitland,

I implemented your approach link, changed accordly and it works properly now.

Thank you for your help.
